core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c

```diff
@@ -691,6 +691,14 @@ pthread_join_wrapper(wasm_exec_env_t exec_env, uint32 thread,
         bh_assert(node->joinable);
         join_ret = 0;
         ret = node->u.ret;
+
+        /* The target thread changes the node's status before calling
+           wasm_cluster_exit_thread to exit, so here its resources may
+           haven't been destroyed yet, we wait enough time to ensure that
+           they are actually destroyed to avoid unexpected behavior. */
+        os_mutex_lock(&exec_env->wait_lock);
+        os_cond_reltimedwait(&exec_env->wait_cond, &exec_env->wait_lock, 1000);
+        os_mutex_unlock(&exec_env->wait_lock);
```

Comment:
isn't this still racy?

Reply:
The `exec_env->wait_lock` is used singly or together with `exec_env->wait_cond`, it won't be occupied by a thread for a long time: when a thread locks `exec_env->wait_lock`, it will unlock it after short time operations, or call os_cond_wait/reltimedwait, note that the lock will be unlocked firstly inside os_cond_wait/reltimedwait. My understanding is that here the current thread should be able to acquire the lock.